### PR TITLE
keys: outline the crdb keyspace layout

### PR DESCRIPTION
**Pull request by irfansharif**
I keep finding myself consulting the combination of
pkg/keys/{keys,docs}.go to remind myself what sits where, and under what
key ranges. Having a top-level reference as such made it easier for
me, and might be useful for others. I also added new documentation
around key addressing, and differences between local and global keys.

I've additionally reordered constant declarations in
pkg/keys/constants.go to more closely follow the physical ordering of
keys. I've grouped sets of prefixes/infixes/suffixes to the closest
approximation of the physical grouping. It's not perfect, some
suffixes are reused across categories for e.g.

Release note: None

### CODE REVIEW

**Comment by irfansharif** [comment]
@sumeerbhola: Since you also mentioned being confused by this area, what else do you think would've been useful to have? (prose, etc.)

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/keys/doc.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBNTPT0waVPWB93wkO:-LyBNTPT0waVPWB93wkP:b-61seq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L20)):*
> ```Go
> // how to decode SQL keys.
> //
> // 1. Overview
> ```

Does this still render as a section in godoc?
___
*[pkg/keys/doc.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBNZJLAo3Op0cw1-Rm:-LyBNZJM6xLYcw5tHuu8:bg51kz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L89)):*
> ```Go
> package keys
> 
> // TODO(irfansharif): Sort the layout below as per the physical representation.
> ```

Are you planning to do this in this PR?
___
*[pkg/keys/doc.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBNfm8AY7yxiu8hLTr:-LyBNfm8AY7yxiu8hLTs:bicaq2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L90)):*
> ```Go
> 
> // TODO(irfansharif): Sort the layout below as per the physical representation.
> func keymap() { //lint:ignore U1000
> ```

nit: It might be cleaner and not require a lint directive if you define this as an array:
```
var keymap = [...]interface{}{
    StoreClusterVersionKey,
    ...
}
```
___
*[pkg/keys/doc.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBO3iQ4Y7hA15d8y8o:-LyBO3iQ4Y7hA15d8y8p:b757w1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L146)):*
> ```Go
> 	// This often makes the choice between the two clear.
> 
> 	// TODO(irfansharif): Fill out keymap for the global keyspace (meta
> ```

Are you planning to do that in this PR?

**Review by nvb — COMMENTED** [review]
I like this idea a lot! We'll have to remember to keep the keymap up to date though. We might want to add another "Note" to `constants.go`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LyBOIem1Cdl0OSvOO49:bleq4l5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

**Review by irfansharif — COMMENTED** [review]
TFTR! PTAL, I took some liberties with the re-ordering and grouping of constants in `constants.go`. (Happy to revert those, but I think they make things a bit clearer still)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LyBmTEW7hzyi6FdI2hP:bwqq577)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/keys/doc.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBNTPT0waVPWB93wkO:-LyBjsbm0K1NaRw-Xhh7:bglfpfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L20)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this still render as a section in godoc?
</blockquote></details>

No, unfortunately not. I figured it was better to have actual symbols in place as opposed to the raw godoc. I've mentioned where to look in the godoc, and put it all in doc.go as a stopgap.
___
*[pkg/keys/doc.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBNZJLAo3Op0cw1-Rm:-LyBk4akD8Jszr-Z-qA7:b4r94mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L89)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are you planning to do this in this PR?
</blockquote></details>

Not originally, but done.
___
*[pkg/keys/doc.go, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBNfm8AY7yxiu8hLTr:-LyBmRAW20Hx8ri94kK0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L90)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: It might be cleaner and not require a lint directive if you define this as an array:
```
var keymap = [...]interface{}{
    StoreClusterVersionKey,
    ...
}
```
</blockquote></details>

Done.
___
*[pkg/keys/doc.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyBO3iQ4Y7hA15d8y8o:-LyBkPOS9B018ROOfEpO:bevftus) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bf7d08b2c4e048347429f6f694dd7345177ff91/pkg/keys/doc.go#L146)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are you planning to do that in this PR?
</blockquote></details>

Done, PTAL.

**Comment by irfansharif** [comment]
I changed it to the array form, but we'd still need a lint directive? We're not using the symbol either way.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
Thanks for doing this!

> Reviewed 2 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LyFut_073hPLFb0Wtzm:bxgavm9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/keys/doc.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFk_aw3WHJalihw4Ch:-LyFk_aw3WHJalihw4Ci:bxqn44x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L49)):*
> ```Go
> //
> // Non-local key types (i.e., meta1, meta2, system, and SQL keys) are
> // collectively referred to as "global" keys. The empty key (/Min) is a special
> ```

I think it would be good to give a 1 line definition of what is a local key and a global key here, even though it may need more refining below. Otherwise this is where the reader starts getting lost. Also some examples of local key types (just like there are examples here of global key types) would be helpful.
___
*[pkg/keys/doc.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFl-7e5hpFg6O7ariA:-LyFl-7e5hpFg6O7ariB:bdvsh2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L57)):*
> ```Go
> // 2. Key addressing
> //
> // The keyspace is divided into contiguous, non-overlapping chunks called
> ```

Should this be "The global keyspace ..."? And does every local key map/address into a key range in the global key space?
___
*[pkg/keys/doc.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFlTTy9-QnKyjLIHWK:-LyFlTTy9-QnKyjLIHWL:b-2icnvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L62)):*
> ```Go
> // the upper bound is exclusive. Any key that begins with /Table/1, like
> // /Table/1/SomePrimaryKeyValue..., would belong to this range. Range boundaries
> // are always global keys.
> ```

"are always global keys" is confusing unless ranges are defined only over the global keyspace.
___
*[pkg/keys/doc.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFmNnHEKNjEdLcLpP5:-LyFmNnHEKNjEdLcLpP6:bij3a6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L69)):*
> ```Go
> // naturally sort into the range [/Min, /System), but its address is /Table/1,
> // so it actually belongs to a range like [/Table1, /Table/2). Note that not
> // every local key has an address.
> ```

Is the address always the start key of a range or can it also lie somewhere inside the range? Related question: is the address of two different local keys that belong to the same range also different, or can the address-key collide?
___
*[pkg/keys/doc.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFmgdP8lHUzdzVGRBg:-LyFmgdP8lHUzdzVGRBh:ba3h7kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L77)):*
> ```
> hold data local to a RocksDB instance
> ```

this suggests that it is not part of the state machine state maintained via raft, but my understanding is that some local keys are replicated. If that is correct, can this phrasing be changed?
___
*[pkg/keys/doc.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFtfRF6u4Z61NCzGcg:-LyFtfRF6u4Z61NCzGch:b-h8q7jy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L95)):*
> ```Go
> 	// range-ID, unreplicated range-ID, range local, and store-local keys.
> 	//
> 	// N.B: The sorting order of the symbols below map to the physical layout.
> ```

And the groups are sorted wrt each other according to the physical layout?
___
*[pkg/keys/doc.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFr6qK4dO-s47bK5Af:-LyFr6qK4dO-s47bK5Ag:b-1vbgkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L99)):*
> ```Go
> 	//
> 	//   1. Replicated range-ID local keys: These store metadata pertaining to a
> 	//   range as a whole. Though they are replicated, they are unaddressable.
> ```

what does "unaddressable" mean in this context? Does it mean one can't call `Addr()` on these `roachpb.Keys`, and if so why not? (I am foggy about the contexts that need to use Addr() -- I would have expected one would have a need to map every local key to its corresponding range).
___
*[pkg/keys/doc.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFqePx-wrEvGvxWsAY:-LyFqePx-wrEvGvxWsAZ:b-g8r1ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L101)):*
> ```Go
> 	//   range as a whole. Though they are replicated, they are unaddressable.
> 	//   Typical examples are MVCC stats and the abort span.
> 	AbortSpanKey,                    // "abc-"
> ```

Is the string in the comment the "suffix"?
I don't understand the prefix, infix, suffix, detail scheme of the key encoding; which key types use what scheme and why? Is there an existing comment somewhere that explains?
___
*[pkg/keys/doc.go, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFt1lv2-vweVC6lVsM:-LyFt1lv2-vweVC6lVsN:bc0i0nh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L126)):*
> ```Go
> 	//   3. Range local keys: These also store metadata that pertains to a range
> 	//   as a whole. They are replicated and addressable. Typical examples are
> 	//   the local range descriptor and transaction records.
> ```

what makes this a "local range descriptor"? Or is this just a "range descriptor" and the "local" was alluding to the fact that all range descriptors are stored under "local keys"?
___
*[pkg/keys/doc.go, line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFuLvg159xZaUTt4us:-LyFuLvg159xZaUTt4ut:bgkx4x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L141)):*
> ```Go
> 	//   4. Store local keys: These contain metadata about an individual store.
> 	//   They are unreplicated and unaddressable. The typical example is the
> 	//   store 'ident' record.
> ```

these also map to specific ranges, yes?

are there any keys in the store that pertains to the node and not to a particular range?

**Review by irfansharif — COMMENTED** [review]
Reworded more things in the top level documentation, specifically around key addressing and the differences between local and global keys. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LyUwJNsEn6VMXV-x3tt:b-fa8lk4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/keys/doc.go, line 49 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFk_aw3WHJalihw4Ch:-LyUVtz8EuVoDXDq1C9K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L49)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think it would be good to give a 1 line definition of what is a local key and a global key here, even though it may need more refining below. Otherwise this is where the reader starts getting lost. Also some examples of local key types (just like there are examples here of global key types) would be helpful.
</blockquote></details>

Done.
___
*[pkg/keys/doc.go, line 57 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFl-7e5hpFg6O7ariA:-LyUXUH41NK3U6BkgLXs:bcylrcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L57)):*
> Should this be "The global keyspace ..."?

No, but agreed that it is confusing. The "local key ranges" are nigh useless. We introduced the "Address" concept to assign local keys a range _other_ than the one they would normally belong to. Added some more clarifying language.

> And does every local key map/address into a key range in the global key space?

No, we mention below that "not every local key has an address". Only range local keys incorporating a range key (start key or transaction key) are addressable (e.g. range metadata and txn records). Range local keys incorporating the Range ID are not (e.g. AbortSpan Entries, and range stats). Added all this as text in explaining key addressing.
___
*[pkg/keys/doc.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFlTTy9-QnKyjLIHWK:-LyUZfwp6AWAKS2je-Pi:b-4qlj6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L62)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

"are always global keys" is confusing unless ranges are defined only over the global keyspace.
</blockquote></details>

Replied above.
___
*[pkg/keys/doc.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFmNnHEKNjEdLcLpP5:-LyUbFhzAkWTJ3RYjBBu:bc8gpym) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L69)):*
I think "address" is bit of a misnomer. The way to understand all this is imagining the function `func resolve(key) rkey` where `rkey` is a "resolved key". Our definition for key ranges is actually over "resolved key ranges". For global keys, the resolved key is simply the key itself. For certain kinds of local keys, the resolved key sits in the global namespace (I'm abusing the definition of ranges, sigh). So calling `keys.Addr(key)` gives us the resolved key. Now, using this understanding for your questions:

> Is the address always the start key of a range or can it also lie somewhere inside the range?

The latter, the "address" of a key should instead be though of as the "resolving of a key", which for certain local keys is simply removal of the local prefix + infix bytes.

> Is the address of two different local keys that belong to the same range also different, or can the address-key collide?

They're different. If the "address-keys" were the same, that would mean the local keys are the same.
___
*[pkg/keys/doc.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFmgdP8lHUzdzVGRBg:-LyUdXLM1n96j14w-xfD:bevftus) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L77)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> ```
> hold data local to a RocksDB instance
> ```

this suggests that it is not part of the state machine state maintained via raft, but my understanding is that some local keys are replicated. If that is correct, can this phrasing be changed?
</blockquote></details>

Done, PTAL.
___
*[pkg/keys/doc.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFtfRF6u4Z61NCzGcg:-LyULtBA7UJ-q0dqBhiV:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L95)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

And the groups are sorted wrt each other according to the physical layout?
</blockquote></details>

Yes.
___
*[pkg/keys/doc.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFr6qK4dO-s47bK5Af:-LyUaSt32PJRmOCSq9dE:bxo4rsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L99)):*
> Does it mean one can't call Addr() on these roachpb.Keys

Correct. 

> and if so why not? 

Added more language around the "Key Addressing" section, and a comment above.
___
*[pkg/keys/doc.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFqePx-wrEvGvxWsAY:-LyURA5v65e7IsDT0fgD:b-harlxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L101)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is the string in the comment the "suffix"?
I don't understand the prefix, infix, suffix, detail scheme of the key encoding; which key types use what scheme and why? Is there an existing comment somewhere that explains?
</blockquote></details>

Added explanations for all this. 	

    // Local keys are constructed using a prefix, an optional infix, and a
    // suffix. The prefix and infix are used to disambiguate between the four
    // types of local keys listed above, and determines inter-group ordering.
    // The string comment next to each symbol below is the suffix pertaining to
    // the corresponding key (and determines intra-group ordering).
    //
    // - RangeID replicated keys all share `LocalRangeIDPrefix` and `LocalRangeIDReplicatedInfix`.
    // - RangeID unreplicated keys all share `LocalRangeIDPrefix` and `localRangeIDUnreplicatedInfix`.
    // - Range local keys all share `LocalRangePrefix`.
    // - Store keys all share `localStorePrefix`.
___
*[pkg/keys/doc.go, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFt1lv2-vweVC6lVsM:-LyU_TevEjWgPeCHZfbH:bln7yp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L126)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what makes this a "local range descriptor"? Or is this just a "range descriptor" and the "local" was alluding to the fact that all range descriptors are stored under "local keys"?
</blockquote></details>

The latter, removed "local" from the phrasing.
___
*[pkg/keys/doc.go, line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyFuLvg159xZaUTt4us:-LyU_Wwp7g1QYloHT_WX:bt5exum) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b58411bee5625660ceab05a37a61ae05b550dc1/pkg/keys/doc.go#L141)):*
> these also map to specific ranges, yes?

No. "store local keys", "range ID replicated", and "range ID unreplicated" keys are all explicitly non-addressable. Check out the implementation of `keys.Addr` to confirm. Only "range local keys" are addressable. Added more clarifying language around the "Key Addressing" section.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
Only some small remaining questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LyVAnJhC6YypEQOOUWS:b-bwvflr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/keys/doc.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV7283C7jxn0somaWH:-LyV7283C7jxn0somaWI:bqpmfnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L87)):*
> ```Go
> // We also have this concept of the "address" for a key. Keys get "resolved"
> // using `keys.Addr`, through which we're able to lookup the range "containing"
> // the key. For global keys, the resolved key is the key itself.
> ```

Presumably even some non-resolvable/addressable keys like the replicated/unreplicated range-id local keys allow for such lookup in that one can recover the range-id and then lookup the range based on the range-id, yes?
And if I am reading correctly these range-id local keys use the range-id in the key, while the "range local keys" use a key within the range bounds -- why this difference? Was it because state written to the latter needs to survive range splits and merges, like for the `TransactionKey`? If yes, does the same need to be true for the `RangeDescriptorKey` -- presumably no since we expect exactly one per range?
___
*[pkg/keys/doc.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV6DAZ6V4glYE_Fmus:-LyV6DAZ6V4glYE_Fmut:bw4a1d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L92)):*
> ```Go
> // ones), the resolved key is obtained by stripping out the local key prefix,
> // suffix, and optional details (refer to `keymap` below to understand how local
> // keys are constructed). This level of indirection was introduced so that we
> ```

Relating to a question in the previous round of comments, doesn't this stripping mean that different range local keys may have identical resolved keys?
And is this resolved key always the start key of the range -- my understanding based on looking at `RangeDescriptorKey()` is that it is the start key? Or when first constructed and written it is the start key but after splits-merges it can be anywhere inside?

**Review by irfansharif — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/keys/doc.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV7283C7jxn0somaWH:-LyVHdrJ5PKJhtvexTCj:bvp8ccb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L87)):*
> Presumably even some non-resolvable/addressable keys like the replicated/unreplicated range-id local keys allow for such lookup in that one can recover the range-id and then lookup the range based on the range-id, yes?

Yes, but at that point, we're decoding the byte layout of the range-id local key and performing look ups based on that. I don't think I quite follow the question here.

> why the difference?

Hm, I'm actually not sure. Presently the split seems to be between whats "addressable" and not (and if not, if it's replicated or not). 
From the comments, "only addressable keys, like range local keys, can be the target of KV operations". But I don't know if this was the only factor considered when introducing this categorization. Perhaps @nvanbenschoten can chime in?
___
*[pkg/keys/doc.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV6DAZ6V4glYE_Fmus:-LyVF47_1cS2V6he6DAj:b76u2r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L92)):*
> Relating to a question in the previous round of comments, doesn't this stripping mean that different range local keys may have identical resolved keys?

I'm not sure I completely understand this question. Taking `RangeDescriptorKey` for example, it's structured as the following:

```
<LocalRangePrefix+key+LocalRangeDescriptorSuffix>
```

The resolved address for this strips out `LocalRangePrefix` and `LocalRangeDescriptorSuffix`, leaving _only_ `key`. `LocalRangePrefix` and `LocalRangeDescriptorSuffix` are shared by all `RangeDescriptorKey`s. When I read "different range descriptor keys", I therefore necessarily think of range descriptor keys with different `key` infixes. And that's what I mean when saying different range descriptor keys resolve to different resolved keys.

The same idea holds true, at the time of writing, for all range local keys. So as I understand it: different range local keys have different resolved keys. Does this perhaps answer your question?

> And is this resolved key always the start key of the range -- my understanding based on looking at RangeDescriptorKey() is that it is the start key? Or when first constructed and written it is the start key but after splits-merges it can be anywhere inside?

The resolved key for `RangeDescriptorKey` is whatever `key` is as diagrammed above. If it happens to be the start key of the range, I believe that's simply an implementation detail for `RangeDescriptorKey`. That is to say, it's not part of the `Addr` API contract. So it can be "anywhere inside" as per your example. My understanding of this is simply that "Addr" provides a logical mapping from one keyspace (addressable local keys + global keys) to another keyspace (set of all resolved keys, which is the global keyspace interleaved with the addressed local keys).

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @nvanbenschoten, and @sumeerbhola)

___
*[pkg/keys/doc.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV6DAZ6V4glYE_Fmus:-LyZJpGw9D3AVQhtf7cv:bvr0nzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L92)):*
> The resolved address for this strips out LocalRangePrefix and LocalRangeDescriptorSuffix, leaving only key. LocalRangePrefix and LocalRangeDescriptorSuffix are shared by all RangeDescriptorKeys. When I read "different range descriptor keys", I therefore necessarily think of range descriptor keys with different key infixes. And that's what I mean when saying different range descriptor keys resolve to different resolved keys.

Yes this clarifies things for `RangeDescriptorKey`.  For `TransactionKey`s that are in the same range, it seems that different ones can resolve to the same resolved key since the `Addr()` strips away the detail too which is what has the txn ID. Is that correct?

**Review by irfansharif — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[pkg/keys/doc.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV6DAZ6V4glYE_Fmus:-LyZk7jS6ovW1nr6ISkc:bvky79k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L92)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> The resolved address for this strips out LocalRangePrefix and LocalRangeDescriptorSuffix, leaving only key. LocalRangePrefix and LocalRangeDescriptorSuffix are shared by all RangeDescriptorKeys. When I read "different range descriptor keys", I therefore necessarily think of range descriptor keys with different key infixes. And that's what I mean when saying different range descriptor keys resolve to different resolved keys.

Yes this clarifies things for `RangeDescriptorKey`.  For `TransactionKey`s that are in the same range, it seems that different ones can resolve to the same resolved key since the `Addr()` strips away the detail too which is what has the txn ID. Is that correct?
</blockquote></details>

Oh, you're right, missed that. I think most of what I understand of our usage of `Addr` is still true: it's there to simply have TransactionKeys (or range local keys more generally) sort into the right range. Grepping through a few usages of `Addr`, we only ever use it for bounds detection ('does the resolved key sit between [start,end)?'). This in turn informs collocation, sorting, etc. but we keep a reference to the original key (with the detail, if available), if needed.

**Review by sumeerbhola — APPROVED** [review]
:lgtm: pending an answer to the question about range-id based lookup

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LydayqnEQiAUyaVKoDQ:bbkcf98)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 2 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-:-LyeiEHY-LpeuhX-Azoy:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/keys/doc.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV7283C7jxn0somaWH:-Lyeh5Uo1fmVFr4RuxMZ:b-y50xjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L87)):*
> Presumably even some non-resolvable/addressable keys like the replicated/unreplicated range-id local keys allow for such lookup in that one can recover the range-id and then lookup the range based on the range-id, yes?

I don't think this is possible. `keys.Addr`, which is called from `verifyKeys` will throw an error on range-ID keys.

> And if I am reading correctly these range-id local keys use the range-id in the key, while the "range local keys" use a key within the range bounds -- why this difference?

The difference between "range-id local keys" and "range local keys" is that range local keys split and merge along range boundaries while range-id local keys don't. This naturally leads to range-id local keys being used to store metadata about a specific Range and range local keys being used to store metadata about specific global keys.
___
*[pkg/keys/doc.go, line 92 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43858#-LyV6DAZ6V4glYE_Fmus:-LyehuiA3WWd3oWrc89h:biymkn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e3472c224e64cef0c4772e6cd98811ed6d4b273/pkg/keys/doc.go#L92)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Oh, you're right, missed that. I think most of what I understand of our usage of `Addr` is still true: it's there to simply have TransactionKeys (or range local keys more generally) sort into the right range. Grepping through a few usages of `Addr`, we only ever use it for bounds detection ('does the resolved key sit between [start,end)?'). This in turn informs collocation, sorting, etc. but we keep a reference to the original key (with the detail, if available), if needed.
</blockquote></details>

Yes, different range local keys can address to the same resolved key.

**Comment by irfansharif** [comment]
TFTRs, this was very instructive.

**Comment by irfansharif** [comment]
> TestLint/TestMisspell: "behavioural" is a misspelling of "behavioral"

🙄

**Comment by irfansharif** [comment]
bors retry

**Pull request merged**
